File: Challenge/May_LeetCoding_Challenge/Week_3:_May_15th–May_21st/230_Kth_Smallest_Element_in_a_BST/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
from typing import List


class Solution:

	def kthSmallest(self, root: TreeNode, k: int) -> int:
		arr: List[int] = []
		self.inorder(root, arr)
		return arr[k - 1]

	def inorder(self, root: TreeNode, arr: List[int]) -> None:
		if root is not None:
			self.inorder(root.left, arr)
			arr.append(root.val)
			self.inorder(root.right, arr)

File: Challenge/May_LeetCoding_Challenge/Week_3:_May_15th–May_21st/230_Kth_Smallest_Element_in_a_BST/solution.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func kthSmallest(root *TreeNode, k int) int {
    order, ans, finished := 0, 0, false
    inorder(root, k, &order, &ans, &finished)
    return ans
}

func inorder(root *TreeNode, k int, order *int, ans *int, finished *bool) {
    if !*finished && root != nil {
        inorder(root.Left, k, order, ans, finished)
        *order++
        if *order == k {
            *ans, *finished = root.Val, true
            return
        }
        inorder(root.Right, k, order, ans, finished)
    }
}